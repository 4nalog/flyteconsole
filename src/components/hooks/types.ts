import {
    Execution,
    NodeExecution,
    PaginatedEntityResponse,
    RequestConfig
} from 'models';

export interface FetchArgs {
    force?: boolean;
}

export interface FetchableData<T> {
    fetch(fetchArgs?: FetchArgs): Promise<T>;
    hasLoaded: boolean;
    lastError: Error | null;
    loading: boolean;
    debugName: string;
    value: T;
}

export interface FetchableExecution {
    fetchable: FetchableData<Execution>;
    terminateExecution(cause: string): Promise<void>;
}

export interface PaginatedFetchableData<T> extends FetchableData<T[]> {
    /** Whether or not a fetch would yield more items. Useful for determining if
     * a "load more" button should be shown
     */
    moreItemsAvailable: boolean;
}

export type FetchFn<T, DataType> = (
    data: DataType,
    currentValue?: T
) => Promise<T>;

export type PaginatedFetchFn<T, DataType = undefined> = (
    data: DataType,
    config: RequestConfig
) => Promise<PaginatedEntityResponse<T>>;

/** Defines parameters for auto-refreshing an entity of type T */
export interface RefreshConfig<T> {
    /** How often (ms) to check the finality of the entity and perform a refresh */
    interval: number;
    /** Implements a boolean check for whether this entity is in a final state
     * and should no longer be refreshed
     */
    valueIsFinal: (value: T) => boolean;
    /** Implements the fetch logic for this entity. Defaults to calling fetch() */
    doRefresh?: (fetchable: FetchableData<T>) => Promise<T>;
}

export type NodeExecutionDictionary = Record<string, NodeExecution>;
