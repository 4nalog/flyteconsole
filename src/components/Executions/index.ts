export * from './constants';
export * from './ExecutionStatusBadge';
export * from './TaskExecutionsList';
export * from './utils';
