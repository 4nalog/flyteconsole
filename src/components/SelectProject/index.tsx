export * from './ProjectList';
export * from './SelectProject';
