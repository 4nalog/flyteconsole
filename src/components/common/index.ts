export * from './ButtonCircularProgress';
export * from './ColoredDot';
export * from './ContentContainer';
export * from './DetailsGroup';
export * from './DetailsPanel';
export * from './DetailsPanelContent';
export * from './DumpJSON';
export * from './ErrorBoundary';
export * from './ExpandableContentLink';
export * from './types';
export * from './LoadingSpinner';
export * from './NewTargetLink';
export * from './NonIdealState';
export * from './SearchableList';
export * from './SectionHeader';
export * from './SectionSubheader';
export * from './ViewHeader';
export * from './WaitForData';
export * from './withRouteParams';
