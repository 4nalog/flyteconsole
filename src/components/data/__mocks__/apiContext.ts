import { APIContextValue, defaultAPIContextValue } from '../apiContext';

export function mockAPIContextValue(
    overrides: Partial<APIContextValue>
): APIContextValue {
    const emptyImpl = Object.keys(defaultAPIContextValue).reduce<
        APIContextValue
    >(
        (out, key) =>
            Object.assign(out, {
                [key]: () => Promise.reject(` ${key} not implemented`)
            }),
        {} as APIContextValue
    );
    return Object.assign(emptyImpl, overrides);
}
