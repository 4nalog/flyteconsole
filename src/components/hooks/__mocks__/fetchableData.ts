import { FetchableData } from '../types';

export function createMockFetchable<T>(value: any): FetchableData<T> {
    return {
        value,
        debugName: '',
        hasLoaded: false,
        fetch: jest.fn(),
        lastError: null,
        loading: false
    };
}

export function loadingFetchable<T>(value: any): FetchableData<T> {
    return { ...createMockFetchable(value), loading: true };
}

export function loadedFetchable<T>(value: any): FetchableData<T> {
    return { ...createMockFetchable(value), hasLoaded: true };
}
