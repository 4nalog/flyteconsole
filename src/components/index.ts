export * from './App';
export * from './common';
export * from './Executions';
export * from './Navigation';
export * from './NotFound';
export * from './SelectProject';
export * from './Tables';
export * from './Workflow';
export * from './WorkflowGraph';
