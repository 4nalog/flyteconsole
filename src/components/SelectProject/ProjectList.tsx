import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import { makeStyles, Theme } from '@material-ui/core/styles';
import { useCommonStyles } from 'components/common/styles';
import { Project } from 'models/Project';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { Routes } from 'routes/routes';

const useStyles = makeStyles((theme: Theme) => ({
    projectCard: {
        textAlign: 'left',
        marginBottom: theme.spacing(2),
        width: theme.spacing(36)
    },
    projectTitle: {
        paddingBottom: 0
    }
}));

export interface ProjectListProps {
    error?: string;
    projects: Project[];
}

const DomainLinks: React.FC<{ project: Project }> = ({ project }) => {
    return (
        <section>
            {project.domains.map(({ id: domainId, name }, idx) => (
                <span key={domainId}>
                    {idx > 0 && <span>&nbsp;|&nbsp;</span>}
                    <Link
                        to={Routes.ProjectDetails.sections.workflows.makeUrl(
                            project.id,
                            domainId
                        )}
                    >
                        {name}
                    </Link>
                </span>
            ))}
        </section>
    );
};

const ProjectCard: React.FC<{ project: Project }> = ({ project }) => {
    const styles = useStyles();
    return (
        <Card className={styles.projectCard}>
            <CardHeader
                className={styles.projectTitle}
                title={project.name}
                titleTypographyProps={{ variant: 'body1' }}
            />
            <CardContent>
                <DomainLinks project={project} />
            </CardContent>
        </Card>
    );
};

/** Displays the available Projects and domains as a list of cards */
export const ProjectList: React.FC<ProjectListProps> = props => {
    const commonStyles = useCommonStyles();
    return (
        <ul className={commonStyles.listUnstyled}>
            {props.projects.map(project => (
                <li key={project.id}>
                    <ProjectCard project={project} />
                </li>
            ))}
        </ul>
    );
};
