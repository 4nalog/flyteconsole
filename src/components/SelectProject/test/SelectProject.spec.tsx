import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { useProjects } from 'components/hooks';
import {
    loadedFetchable,
    loadingFetchable
} from 'components/hooks/__mocks__/fetchableData';
import { Project } from 'models';
import { createMockProjects } from 'models/__mocks__/projectData';

import { SelectProject } from '../SelectProject';

// Skipping full render of the ProjectList component because it has its own snapshot
jest.mock('../ProjectList');
jest.mock('components/hooks');

// TODO: Enzyme doesn't yet have support for hooks, and the timeline isn't looking like any time soon. The current best option is to switch to using react-testing-library instead.
describe.skip('SelectProject', () => {
    const useProjectsMock = useProjects as jest.Mock<
        ReturnType<typeof useProjects>
    >;
    const renderSnapshot = () => renderer.create(<SelectProject />).toJSON();

    describe('while loading', () => {
        beforeEach(() => {
            useProjectsMock.mockReturnValue(loadingFetchable([]));
        });
        it('should match the known snapshot', () => {
            expect(renderSnapshot()).toMatchSnapshot();
        });
    });

    describe('After successfully loading', () => {
        let projects: Project[];
        beforeEach(() => {
            projects = createMockProjects();
            useProjectsMock.mockReturnValue(loadedFetchable(projects));
        });

        it('should match the known snapshot', () => {
            expect(renderSnapshot()).toMatchSnapshot();
        });
    });
});
