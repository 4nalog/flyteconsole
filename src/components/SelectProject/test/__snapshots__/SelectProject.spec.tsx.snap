// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectProject After successfully loading should match the known snapshot 1`] = `
<div
  className="makeStyles-container-88"
>
  <div
    className="makeStyles-container-92"
  >
    <div
      style={
        Object {
          "fontSize": 64,
        }
      }
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-94 MuiSvgIcon-colorAction-97 MuiSvgIcon-fontSizeInherit-100"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <h5
      className="MuiTypography-root-17 MuiTypography-h5-26 makeStyles-section-93"
    >
      Unexpected error
    </h5>
    <p
      className="MuiTypography-root-17 MuiTypography-body1-19 makeStyles-section-93"
    >
      <div
        className="makeStyles-messageContainer-91"
      >
        <p>
          The error we received was:
        </p>
        <div
          className="MuiPaper-root-104 MuiPaper-elevation1-107 MuiPaper-rounded-105 MuiCard-root-103 makeStyles-errorContainer-90"
        >
          <div
            className="MuiCardContent-root-131 makeStyles-codeBlock-78"
          >
            Invariant Violation: mockConstructor(...): Nothing was returned from render. This usually means a return statement is missing. Or, to render nothing, return null.
          </div>
        </div>
        <p>
          There may be additional information in the browser console.
        </p>
      </div>
    </p>
  </div>
</div>
`;

exports[`SelectProject while loading should match the known snapshot 1`] = `
<div
  className="makeStyles-container-4 large"
>
  <div
    className="MuiCircularProgress-root-5 MuiCircularProgress-colorPrimary-8 MuiCircularProgress-indeterminate-7"
    role="progressbar"
    style={
      Object {
        "height": 96,
        "width": 96,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg-10"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle-11 MuiCircularProgress-circleIndeterminate-13"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;
